### ISSUE TITLE
cf_instance_id is a Compound Attribute of app_guid and cf_instance_number / Possible Cardinality Explosion?

### ISSUE THREAD

**Issue by dmarmugi** [promregator/promregator#92]
## Environment
Promregator v0.5.6
macOS 10.14.3
Jar
cf api 2.114.0

## Summary / Problem Statement
Apparent cardinality explosion from `cf_instance_id` label. The `cf_instance_id` seems to be a concatenation of `<app guid>:<instance number>`. 

## Observed Behavior
```
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557:5",cf_instance_number="5",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557:3",cf_instance_number="3",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557:1",cf_instance_number="1",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557:7",cf_instance_number="7",} 0.0
```

## Expected Behavior
Either the `:<instance number>` should be lopped off when generating the metrics, or the label should be suppressed, to reduce the cardinality and make querying on the field usable.

**Comment by dmarmugi** [comment]
So to clarify: it's not an earth-shattering cardinality explosion since there's an upper-limit, but as-is, it's not much more helpful than just querying `org + space + app`

**Comment by eaglerainbow** [comment]
Hi dmarmugi,

Welcome to Promregator and thank you for providing feedback!

I am trying to understand your issue, but I have to admit that I couldn't really get your point so far. I am not able to understand which would be the misbehavior based on the example and the statements you made. I'd like to understand where our two threads of mind start to diverge. For that let me explain what I understood from your example above:

```
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557:5",cf_instance_number="5",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557:3",cf_instance_number="3",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557:1",cf_instance_number="1",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557:7",cf_instance_number="7",} 0.0
```

From the metrics (and their labels) I read the following:

* You are looking at the `promregator_up` metrics, which indicate whether a certain cell / app *instance* is up and could be scraped by Promregator or not.
* All metrics provide value `0.0`, which means that the target could not be scraped.
* I see that there is at least one line with `cf_instance_number="7"` which means that there must have been a point in time where the same application had 8 cells running.
* The fact that they are reported means that they had been up some time before. Apparently, that does not hold true any longer.
* It is to be noted that the reporting here is on the granularity of the *cell* and not on the level of an application, as a single application may consist of one or multiple cells, in which the deployed content is executed. 

You stated that

> Apparent cardinality explosion from cf_instance_id label. The cf_instance_id seems to be a concatenation of <app guid>:<instance number>

Well, it is true that the label `cf_instance_id` is just the concatenation of `<app guid>:<instance number>`. Thus, it would be valid to say that the label `cf_instance_id` is redundant (which was done by purpose, see below)...

However, I can't make the sense out of the following requirement:

> Either the `:<instance number>` should be lopped off when generating the metrics, or the label should be suppressed, to reduce the cardinality

If we dropped the `<instance number>` from `cf_instance_id`, then the value would be equal to the <app_guid>. However, that would not change anything for the number of records returned:

```
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557",cf_instance_number="5",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557",cf_instance_number="3",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557",cf_instance_number="1",} 0.0
promregator_up{org_name="<redacted org>",space_name="<redacted space>",app_name="<redacted app>",cf_instance_id="e22f677d-c261-4a6b-83a2-a3b567511557",cf_instance_number="7",} 0.0
```
The cardinality still remains the same, as `cf_instance_number` still will be there - and has to stay there, as reporting of metrics availability must happen for each cell (if - let's say - cell 3 was down, but cell 5 is still running, then you would want to be able to know that for your monitoring...).

Also suppressing the label `cf_instance_number` completely (which isn't a bright idea, as some consumers use the key `org` + `space` + `app` + `cf_instance_number` and some others use `app_guid:cf_instance_number` to identify a certain cell uniquely) would not change anything: 
* `org_name`, `space_name` and `app_name` will identify an application on the platform uniquely (for the sake of completeness: at a given point in time).
* `cf_instance_number` still will be in place and thus no compression of cardinality would take place either (which - as stated above - you also would not want to). 

So, in short, I would not know how your proposal of changing how the labels were created would reduce the cardinality in first place. On the other hand, I also can't see why you would want to change the cardinality in first place: Reporting of `promregator_up` should indicate, which *cells* are up or down....


Or were you looking for the `up` metric, which Promregator also exposes to document whether *Promregator itself* is up and running?

Please let me know where our understanding starts to diverge. Whether or not we have a bug in Promregator or not, I consider this issue reported here at least as an indicator that there might be something wrong with documentation...

**Comment by dmarmugi** [comment]
Hi @eaglerainbow, thanks for your attention and the detailed response. 
To first clarify a few points: 
* The issue template lends itself to filing something affirmatively as a bug/problem, but I was more just hoping for a discussion
* The example was just a snip from a whole page full of metrics, just meant to highlight the construction of the `cf_instance_id` and its relation to the other labels. 

And to the issue itself: 
As you say, there is no cardinality explosion, because `cf_instance_id` is tightly coupled with the `cf_instance_number` label. It's this tight coupling that makes it seem redundant to me. 

You say: 
> ...some consumers use the key org + space + app + cf_instance_number and some others use app_guid:cf_instance_number to identify a certain cell uniquely

I guess my question is: Why offer consumers 2 ways to do the same thing? Is `cf_instance_id` more useful in some cases than having `app_guid` and `instance_number` as separate dimensions? I may just be ignorant of valid use cases, it just seems un-Prometheus-y to offer a compound label, as you lose any ability to aggregate on it. 

Thanks,
-David

**Comment by dmarmugi** [comment]
That makes sense: backwards compatibility is important. Thank you for that explanation.

As to the app_guid label: 
> Well, based on previous experience, there are no use cases known to me, where you would want to aggregate over either `app_guid` ... alone:

I _thought_ app guids changed, but I wasn't sure until you confirmed it. The case that I had in mind was to enable charting of the canary or blue/green deploy: Something like `count(count(promregator_up{org_name="$org", space_name="$space", app_name="$app", app_guid=~".*"}) by (cf_instance_number))` could show you as a new app version comes online and the old one scales down. You could do similar things to correlate error rates, etc., to app deployments. 

So maybe this ticket is actually about asking for an `app_guid` label :)

**Comment by eaglerainbow** [comment]
> `count(count(promregator_up{org_name="$org", space_name="$space", app_name="$app", app_guid=~".*"}) by (cf_instance_number))`

Well... Already that you write `app_guid=~".*"` in your query tells me, that you don't really need that label 😀  ... And if you really prefer to have such a condition, why can't you use 
```
cf_instance_id=~".*:.*"
``` 

or something similar?

My gut-feeling on this tells me that in case of filtering you'll always get it handled somehow. The only situation in which I suspect that the label really might get necessary is, if you want to group by it. However, grouping usually implies that you show it in the graph - in one way or the other. And these GUIDs aren't really photogenic.

> So maybe this ticket is actually about asking for an app_guid label :)

Please don't get me wrong in assuming that I would not want to provide you such a label, but so far I did not really stumble over a reason which was good enough - and your case couldn't convince me either so far, yet. Sorry 😬 But please keep trying...!

BTW: There is one more point to this discussion. Would the label better be called `app_guid` or `app_id`? 

The CF API uses `app_id`, but what's inside is an `app_guid`. That's a clear "pro `app_id`" for me.
However, if I got started with Promregator and wanted to see "from which app the data is coming", intuitively I would add `app_id` (if existing) to the drilldown - instead of `app_name` (which is most likely what I really wanted to have for a start). In contrast, if there was only `app_name` and `app_guid` to choose from, I think I'd try `app_name` first...

**Comment by eaglerainbow** [comment]
Given that there are no new arguments to this discussion for more than 4 weeks, I would suggest to close this issue here.
If anyone still has a argument to bring forward, please don't hesitate and comment! We can re-open this issue here at any time.
